File: lecture_012(Sorting)/l_012_sortdates.cpp
#include<iostream>
#include<vector>
#include<string>
using namespace std;

void display(vector<string> &arr) {
    for(string r: arr) {
        cout<<r<<endl;
    }
    cout<<endl;
}
void countsort(vector<string> &arr, int div, int mod, int range) {
    vector<int> farr(range,0);
    	
    for(int i=0; i<arr.size(); i++) {
        int idx = stoi(arr[i],10)/div % mod;
        farr[idx]++;
    }
    for(int i=1; i<farr.size(); i++) {
        farr[i] = farr[i] + farr[i-1];
    }
    vector<string> ans(arr.size(),0);

    for(int i=arr.size()-1; i>=0; i--) {
        int val = stoi(arr[i],10)/div % mod;
        int pos = farr[val] - 1;
        ans[pos] = arr[i]; 
        farr[stoi(arr[i],10)/div % mod]--; 
    }

    for(int i=0; i<arr.size(); i++) {
        arr[i] = ans[i];
    }
}
void sortdates(vector<string> &arr) {
    countsort(arr,1000000,100,32);     //days
    countsort(arr,10000,100,13);       //months
    countsort(arr,1,10000,2501);       //years
}

int main() {
    int n; cin>>n;
    vector<string> srr;
    string str;
    for(int i=0; i<n; i++) {
        cin>>str;
        srr.push_back(str);
    }
    sortdates(srr);
    cout<<endl;
    display(srr);

    return 0;
}

File: lecture_012(Sorting)/l_012_sortdates.java
import java.util.Scanner;
public class l_012_sortdates {
    public static Scanner scn = new Scanner(System.in);
    public static void main(String[] args) {
        //int n = scn.nextInt();                    //(wrong)
        int n = Integer.parseInt(scn.nextLine());   //(right)
        String[] srr = new String[n];
        for(int i=0; i<n; i++) {
            srr[i] = scn.nextLine();
        }
        sortDates(srr);
        System.out.println();
        display(srr);
    }
    public static void display(String[] arr) {
    	for(String r: arr) {
    		System.out.println(r);
    	}
    	System.out.println();
    }
    public static void countSort(String[] arr, int div, int mod, int range) {
    	int[] farr = new int[range];
    	
    	for(int i=0; i<arr.length; i++) {
            int idx = Integer.parseInt(arr[i],10)/div % mod;
    		farr[idx]++;
    	}
    	for(int i=1; i<farr.length; i++) {
    		farr[i] = farr[i] + farr[i-1];
    	}
    	String[] ans = new String[arr.length];
	
    	for(int i=arr.length-1; i>=0; i--) {
            int val = Integer.parseInt(arr[i],10)/div % mod;
    		int pos = farr[val] - 1;
    		ans[pos] = arr[i]; 
    		farr[Integer.parseInt(arr[i],10)/div % mod]--; 
    	}

        for(int i=0; i<arr.length; i++) {
    		arr[i] = ans[i];
        }
    }
    public static void sortDates(String[] arr) {
        countSort(arr,1000000,100,32);    //days
        countSort(arr,10000,100,13);      //months
        countSort(arr,1,10000,2501);      //years
    }  
}